### PR TITLE
smite-ir: Add unit tests for `OperationParamMutator`

### PR DESCRIPTION
**Pull request by Chand-ra**
This PR is a follow-up to #44, adding targeted, deterministic unit tests for the `OperationParamMutator`'s helper functions. 

Current tests for this mutator rely primarily on top-level integration loops. Directly testing the helper functions with seeded RNGs guarantees that every individual mutation branch is structurally sound and prevents future regressions (like off-by-one errors).

### CODE REVIEW

File: smite-ir/Cargo.toml [hunk 1]
```diff
@@ -9,7 +9,7 @@ workspace = true
 
 [dependencies]
 smite = { path = "../smite" }
-rand = { version = "0.10", default-features = false }
+rand = { version = "0.10", default-features = false, features = ["std_rng"] }
```
Comment (Chand-ra) [inline smite-ir/Cargo.toml]: `default-features` includes `std_rng` but it's explicitly excluded here, so unsure if this is okay to do. I could add a simple PRNG in the tests module to seed the tests instead, but that'd be writing boilerplate.

File: smite-ir/src/mutators/operation_param.rs [hunk 2]
```diff
@@ -323,3 +323,131 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
 fn interesting_u64(rng: &mut impl Rng) -> u64 {
     INTERESTING_U64[rng.random_range(0..INTERESTING_U64.len())]
 }
+
+#[cfg(test)]
+mod internal_operation_param_mutator_tests {
```
Comment (ekzyis) [inline smite-ir/src/mutators/operation_param.rs]: Any reason why not use `mod tests` here?

File: smite-ir/Cargo.toml (hunk 1 above)
Comment (ekzyis) [reply smite-ir/Cargo.toml]: Can we use `SmallRng` instead of `StdRng` like in smite-ir/src/tests.rs?

File: smite-ir/src/mutators/operation_param.rs [hunk 3]
```diff
@@ -323,3 +323,131 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
+    use super::*;
+    use rand::{SeedableRng, rngs::StdRng};
+
+    #[test]
+    fn test_shuffle_subrange_preserves_elements() {
+        let mut rng = StdRng::seed_from_u64(42);
```
Comment (ekzyis) [inline smite-ir/src/mutators/operation_param.rs]: nit: the tests in smite-ir/src/tests.rs always use 0 as the seed

File: smite-ir/src/mutators/operation_param.rs [hunk 4]
```diff
@@ -323,3 +323,131 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
+        let mut rng = StdRng::seed_from_u64(555);
+
+        // Find a field that has NO other fields with the same output type
+        let unique_field = AcceptChannelField::ALL.iter().find(|&&f| {
+            AcceptChannelField::ALL
+                .iter()
+                .filter(|&&other| other.output_type() == f.output_type())
+                .count()
+                == 1
+        });
```
Comment (ekzyis) [inline smite-ir/src/mutators/operation_param.rs]: nit: I don't expect changes to `AcceptChannelField`, so I think we could hard-code the field here to improve readability.

File: smite-ir/src/mutators/operation_param.rs [hunk 5]
```diff
@@ -323,3 +323,131 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
     INTERESTING_U64[rng.random_range(0..INTERESTING_U64.len())]
 }
+
+#[cfg(test)]
+mod internal_operation_param_mutator_tests {
+    use super::*;
+    use rand::{SeedableRng, rngs::StdRng};
+
+    #[test]
+    fn test_shuffle_subrange_preserves_elements() {
```
Comment (ekzyis) [inline smite-ir/src/mutators/operation_param.rs]: nit: I think the following tests the same while being slightly more readable + it tests that `shuffle_subrange` does something when there's no edge case, which seems to be missing as a test:

```diff
diff --git a/smite-ir/src/mutators/operation_param.rs b/smite-ir/src/mutators/operation_param.rs
index 5bcc2cb..0343998 100644
--- a/smite-ir/src/mutators/operation_param.rs
+++ b/smite-ir/src/mutators/operation_param.rs
@@ -332,24 +332,15 @@ mod internal_operation_param_mutator_tests {
     #[test]
     fn test_shuffle_subrange_preserves_elements() {
         let mut rng = StdRng::seed_from_u64(42);
+        let sorted = vec![1, 2, 3, 4, 5, 6, 7, 8];
+        let mut shuffled = sorted.clone();

-        let mut data = vec![1, 2, 3, 4, 5, 6, 7, 8];
-        let original_len = data.len();
+        shuffle_subrange(&mut shuffled, &mut rng);

-        shuffle_subrange(&mut data, &mut rng);
-
-        assert_eq!(
-            data.len(),
-            original_len,
-            "Shuffle must not change vector length"
-        );
-
-        let mut sorted_data = data.clone();
-        sorted_data.sort_unstable();
-
-        let expected = vec![1, 2, 3, 4, 5, 6, 7, 8];
+        assert_ne!(shuffled, sorted, "Shuffle must shuffle elements");
+        shuffled.sort_unstable();
         assert_eq!(
-            sorted_data, expected,
+            shuffled, sorted,
             "Shuffle must preserve all original elements"
         );
     }
```

File: smite-ir/Cargo.toml
Comment (ekzyis) [inline smite-ir/Cargo.toml:1]: When I run `cargo test -p smite-ir`, it updates Cargo.lock. Could it be that you forgot to push these changes? But I think the changes to Cargo.lock aren't needed if we don't enable `std_rng`, see other comment.

File: smite-ir/Cargo.toml (hunk 1 above)
Comment (Chand-ra) [reply smite-ir/Cargo.toml]: A lot better, thanks.

File: smite-ir/src/mutators/operation_param.rs (hunk 2 above)
Comment (Chand-ra) [reply smite-ir/src/mutators/operation_param.rs]: Wanted to make the name more specific, probably better to go with plain old 'test' though.

File: smite-ir/Cargo.toml
Comment (Chand-ra) [reply smite-ir/Cargo.toml:1]: Done.

File: smite-ir/src/mutators/operation_param.rs (hunk 5 above)
Comment (Chand-ra) [reply smite-ir/src/mutators/operation_param.rs]: Does read better, thanks!

**Review by ekzyis — COMMENTED** [review]
GitHub is having issues ([again](https://news.ycombinator.com/item?id=48034587)). Couldn't create review comments for individual lines, sorry.

L340:

```diff
-        assert_ne!(shuffled, sorted, "Shuffle must shuffle elements");
+        assert_ne!(shuffled, sorted, "shuffle_subrange must shuffle elements");
```

or

```diff
-        assert_ne!(shuffled, sorted, "Shuffle must shuffle elements");
+        assert_ne!(shuffled, sorted, "must shuffle elements");
```

---

L344: same suggestion

---

L424: I think "field that has NO other fields" is confusing

---

L425:

`u32` is slightly inaccurate, block heights are their own type:

```diff
-        // MinimumDepth is the only field with a `u32` output type.
+        // MinimumDepth is the only field with a `BlockHeight` output type.
```

File: smite-ir/src/mutators/operation_param.rs [hunk 6]
```diff
@@ -380,3 +380,115 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
     INTERESTING_U64[rng.random_range(0..INTERESTING_U64.len())]
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::{SeedableRng, rngs::SmallRng};
+
+    #[test]
+    fn test_shuffle_subrange_preserves_elements() {
```
Comment (morehouse) [inline smite-ir/src/mutators/operation_param.rs]: Nit: follow codebase naming convention


```suggestion
    fn shuffle_subrange_preserves_elements() {
```

File: smite-ir/src/mutators/operation_param.rs [hunk 7]
```diff
@@ -380,3 +380,115 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
+        let mut shuffled = sorted.clone();
+
+        shuffle_subrange(&mut shuffled, &mut rng);
+
+        assert_ne!(shuffled, sorted, "shuffle_subrange must shuffle elements");
+        shuffled.sort_unstable();
+        assert_eq!(
+            shuffled, sorted,
+            "shuffle_subrange must preserve all original elements"
+        );
```
Comment (morehouse) [inline smite-ir/src/mutators/operation_param.rs]: This test would be stronger if we looped for many iterations.

File: smite-ir/src/mutators/operation_param.rs [hunk 8]
```diff
@@ -380,3 +380,115 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
+    }
+
+    #[test]
+    fn test_shuffle_subrange_empty_and_single() {
```
Comment (morehouse) [inline smite-ir/src/mutators/operation_param.rs]: Nit: follow codebase naming convention


```suggestion
    fn shuffle_subrange_empty_and_single() {
```

File: smite-ir/src/mutators/operation_param.rs [hunk 9]
```diff
@@ -380,3 +380,115 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
+        shuffle_subrange(&mut empty, &mut rng);
+        assert!(empty.is_empty());
+
+        let mut single = vec![0xFF];
+        shuffle_subrange(&mut single, &mut rng);
+        assert_eq!(single, vec![0xFF]);
+    }
+
+    #[test]
+    fn test_mutate_bytes_respects_max_size() {
```
Comment (morehouse) [inline smite-ir/src/mutators/operation_param.rs]: This is already covered by `param_mutator_caps_byte_length`.

File: smite-ir/src/mutators/operation_param.rs [hunk 10]
```diff
@@ -380,3 +380,115 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
+            assert!(
+                data.len() <= MAX_MESSAGE_SIZE,
+                "mutate_bytes exceeded MAX_MESSAGE_SIZE. Current len: {}",
+                data.len()
+            );
+        }
+    }
+
+    #[test]
+    fn test_mutate_bytes_handles_empty_input() {
```
Comment (morehouse) [inline smite-ir/src/mutators/operation_param.rs]: This test is almost useless.  It runs one time while `mutate_bytes` has 9 randomly selected arms that could execute.  It's also likely already tested in `param_mutator_caps_byte_length`.

File: smite-ir/src/mutators/operation_param.rs [hunk 11]
```diff
@@ -380,3 +380,115 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
+
+        for _ in 0..500 {
+            mutate_fixed_bytes(&mut bytes, &mut rng);
+        }
+        // Ensure the array actually mutated from all zeros
+        assert_ne!(bytes, [0x00; 32], "Array should have been mutated");
+    }
+
+    #[test]
+    fn test_mutate_extract_field_preserves_type() {
```
Comment (morehouse) [inline smite-ir/src/mutators/operation_param.rs]: This is already covered by `param_mutator_preserves_extract_field_type`.

File: smite-ir/src/mutators/operation_param.rs [hunk 12]
```diff
@@ -380,3 +380,115 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
+            assert_eq!(
+                field.output_type(),
+                original_type,
+                "Output type must be preserved"
+            );
+        }
+    }
+
+    #[test]
+    fn test_mutate_extract_field_handles_unique_types() {
```
Comment (morehouse) [inline smite-ir/src/mutators/operation_param.rs]: This is already covered by `param_mutator_returns_false_for_singleton_extract_field`.

**Review by morehouse — CHANGES_REQUESTED** [review]
Most of these tests are already covered in the public API tests.  I think we should drop all the tests except the `shuffle_subrange` ones that actually provide some benefit.

File: smite-ir/src/mutators/operation_param.rs [hunk 13]
```diff
@@ -390,3 +380,38 @@ fn interesting_u32(rng: &mut impl Rng) -> u32 {
...
+        let mut shuffled = sorted.clone();
+
+        for _ in 0..100 {
+            shuffle_subrange(&mut shuffled, &mut rng);
+            shuffled.sort_unstable();
+            assert_eq!(
+                shuffled, sorted,
+                "shuffle_subrange must preserve all original elements"
+            );
+        }
```
Comment (morehouse) [inline smite-ir/src/mutators/operation_param.rs:414]: I think we also should test that shuffle generally does something.


```suggestion
        let mut shuffled = sorted.clone();
        let mut diff_count = 0;

        for _ in 0..100 {
            shuffle_subrange(&mut shuffled, &mut rng);
            if shuffled != sorted {
                diff_count += 1;
            }
            shuffled.sort_unstable();
            assert_eq!(
                shuffled, sorted,
                "shuffle_subrange must preserve all original elements"
            );
        }
        assert(diff_count >= 70, "shuffle_subrange frequently does nothing");
```

Comment (Chand-ra) [reply smite-ir/src/mutators/operation_param.rs:414]: Doesn't look quite right to me.

Since our RNG is deterministic, `diff_count` will evaluate to the exact same number every time for a given seed, at which point we're more so testing the statistical distribution of the `rand` crate rather than the code's logic.

I believe it's better to test that `shuffle_subrange` mutates the input _at least once_. This way we can be sure the function isn't broken and just returning the identical array all the time.

**Comment by Chand-ra** [comment]
Rebased on top of the latest master to resolve a merge conflict.

File: smite-ir/src/mutators/operation_param.rs (hunk 13 above)
Comment (morehouse) [reply smite-ir/src/mutators/operation_param.rs:414]: The idea was to test that there isn't some unexpected bias towards doing nothing (either from our mutator implementation or the RNG), while also reducing the test flakiness in case later code changes cause slightly different mutations to occur.

But either way this is a strict improvement in testing coverage.

**Review by morehouse — APPROVED** [review]
